jgrapht-core/src/main/java/org/jgrapht/alg/spanning/AhujaOrlinSharmaCapacitatedMinimumSpanningTree.java

```diff
@@ -0,0 +1,1336 @@
+         */
+        private void updateSingleNode(
+                SolutionRepresentation currentSolution,
+                Map<V, Pair<Set<V>, Double>> subtrees,
+                Set<V> tabuList,
+                int label,
+                double oldWeight,
+                Set<V> modifiableSet,
+                Pair<Integer, ImprovementGraphVertexType> pseudoVertex,
+                V v1,
+                Pair<Integer, ImprovementGraphVertexType> vertexOfV1Single
+        ) {
+            double newCapacity, newWeight;
+            SpanningTreeAlgorithm.SpanningTree<E> spanningTree;
+
+            // add v1 to the set for MST calculations
+            modifiableSet.add(v1);
+
+            /*
+             * Adding of edges for v1 vertex replacing an object in v2.
+             * We need to considers this only if vertex operations should be used.
+             */
+            if (useVertexOperation) {
+                for (V v2 : currentSolution.getPartitionSet(label)) {
+
+                    if (v2.equals(root)) {
+                        throw new IllegalStateException("The root is in the partition. This is a bug.");
+                    }
+
+                    if (tabuList.contains(v2)) {
+                        continue;
+                    }
+
+                    /*
+                     * edge for v1 vertex replacing v2 vertex
+                     */
+                    modifiableSet.remove(v2);
+                    spanningTree = new PrimMinimumSpanningTree<>(new AsSubgraph<>(graph, modifiableSet, graph.edgeSet())).getSpanningTree();
+                    if (spanningTree.getEdges().size() == modifiableSet.size() - 1) {
+                        newCapacity = calculateMaximumDemandOfSubtrees(modifiableSet, spanningTree, currentSolution.getPartitionWeight(label) + demands.get(v1) - demands.get(v2));
+                        newWeight = spanningTree.getWeight();
+                    } else {
+                        newCapacity = Double.NaN;
+                        newWeight = Double.NaN;
+                    }
+                    updateImprovementGraphEdge(vertexOfV1Single, Pair.of(initialVertexMapping.get(v2), ImprovementGraphVertexType.SINGLE), newCapacity, newWeight - oldWeight);
+                    modifiableSet.add(v2);
+                    // end edge for v1 vertex replacing v2 vertex
+
+                    /*
+                     * edge for v1 vertex replacing v2 subtree
+                     * If the subtree of v2 contains only the vertex itself and both operations are used, it is the same operation as moving v2 as vertex. Thus, do not add edges.
+                     */
+                    if (useSubtreeOperation) {
+                        if (subtrees.get(v2).getFirst().size() > 1) {
+                            modifiableSet.removeAll(subtrees.get(v2).getFirst());
+                            spanningTree = new PrimMinimumSpanningTree<>(new AsSubgraph<>(graph, modifiableSet, graph.edgeSet())).getSpanningTree();
```

Comment:
Our current implementation of `AsSubgraph` is a little bit inefficient. Just saying in order to check whether it has a large impact in running time.

Reply:
Is there any other possibility instead using `AsSubgraph`? Copying the graph is definitly not a solution, hence, `AsSubgraph` has to be improved or I have to use a implementation used only for this purpose, if I consider this correctly.